falcon/cyutil/uri.pyx

```diff
@@ -0,0 +1,299 @@
+# Copyright 2019 by Vytautas Liuolia.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from cpython.mem cimport PyMem_Malloc, PyMem_Free
+from libc.string cimport memcpy
+
+
+cdef list build_hex_table():
+    cdef list result = [-1] * 0x10000
+    for ch1 in '0123456789abcdefABCDEF':
+        for ch2 in '0123456789abcdefABCDEF':
+            try:
+                result[(ord(ch1) << 8) | ord(ch2)] = int(ch1 + ch2, 16)
+            except ValueError:
+                pass
+
+    return result
+
+
+# PERF(vytas): Cache hex characters lookup table
+cdef int[0x10000] HEX_CHARS
+HEX_CHARS[:] = build_hex_table()
+
+# PERF(vytas): Cache an empty string object.
+cdef EMPTY_STRING = u''
+
+
+cdef inline int cy_decode_hex(unsigned char nibble1, unsigned char nibble2):
+    return HEX_CHARS[(nibble1 << 8) | nibble2]
+
+
+cdef unicode cy_decode_no_plus(unsigned char* data, Py_ssize_t start,
+                               Py_ssize_t end, Py_ssize_t encoded_start):
+    if encoded_start < 0:
+        return data[start:end].decode()
+
+    cdef unsigned char* result
+    cdef Py_ssize_t src_start = start
+    cdef Py_ssize_t dst_start = 0
+    cdef Py_ssize_t pos
+    cdef int decoded
+
+    result = <unsigned char*> PyMem_Malloc(end - start)
+    if not result:
+        raise MemoryError()
+
+    try:
+        for pos in range(encoded_start, end):
+            if data[pos] != b'%':
```

Comment:
Is the "no plus" specialization worth the performance increase? If so, it may be a good idea to add some inline comments that point out the differences so that if these functions are ever modified, it makes it more obvious what the developer has to do to specialize  ``cy_decode()``.

I'm all for violating DRY to get a significant speed boost as long as we are careful about the maintenance burden this might introduce _and_ we have comprehensive tests for all paths and a wide variety of inputs. That being said, I can't imagine this code changing that much over time.

Reply:
No, I don't think it's worth the performance increase (if any). I think it was actually the first version that I implemented, and then I had realized I also need to support the default "unquote plus" case too. Human workflow artefact, one can say :thinking:
Simple boolean switches and prediction branching check-ups are dirty fast on modern CPUs, particularly on non-RISC ones thanks to parallel execution pipelines. Sometimes even [too smart for their own good :slightly_smiling_face:](https://en.wikipedia.org/wiki/Spectre_(security_vulnerability)). I think given our default is to unquote `+`-es, I'll be able to DRY this without any noticeable impact.